rfcs/20190506-filesystem-plugin-modular-tensorflow.md

```diff
@@ -0,0 +1,1082 @@
+
+A major module of the modular TensorFlow involves all operations regarding files
+and the filesystem. Saving models, checkpointing, reading files as input to
+models (i.e. images, audio, etc.), and many other file manipulations are all
+tasks that TensorFlow has to support.
+
+While users can use the underlying operating system’s API (such as calling fopen
+and friends on Posix environments, for example), it is definitely better to have
+a cross-platform, cross-environment API that can be used everywhere. TensorFlow
+provides such an API at the moment but it results in compiling code for all
+known filesystems in the final binary. We propose a new design where all
+filesystems are implemented as plugins and only the required ones are added at
+runtime. To do so, we will first need to design a C API for the filesystems and
+then integrate this with the rest of the modularization effort.
+
+In a nutshell, the proposed API uses several data structures of function
+pointers to mimic the current class-based API. Every plugin will fill in the
+appropriate fields of these data structures during registration. When doing a
+filesystem operation, core TensorFlow identifies the plugin/filesystem needed to
+implement the operation, reads the function pointer from the corresponding table
+and executes it. This way, we are able to offer support for all kinds of files,
+including memory mapped files, remote files, compressed files or filesystem with
+specific rules.
+
+The rest of the document presents a set of requirements that this design tries
+to satisfy, a (somewhat very detailed) overview of the existing implementation,
+the proposed changes and estimates on the amount of work needed to integrate
+this design with TensorFlow code.
+
+## Requirements
+
+Moving to a modular design adds several new requirements to the design, as we
+want to prevent code from one plugin to interfere with code from another plugin
+or the core TensorFlow. As such, here are a few requirements that we think are
+needed:
+
+1. Once a plugin is registered, the function pointer to plugin provided
+   implementation should never change. Since unloading shared libraries after
+   code from them has been run can lead to bad behavior (resource leaks, copies
+   of pointers to functions within the library, object size mismatches, etc.) we
+   will not attempt to deregister/unload plugins.
+
+1. Interface between TensorFlow and plugins must be as simple as possible as
+   well as thoroughly tested. It is better to localize errors in either
+   TensorFlow core or the plugin providing the filesystem functionality than in
+   the glue code between the two.
+
+1. Plugin registration should support mandatory metadata such as version
+   numbers, checks for backwards and forwards compatibility as well as optional
+   fields such as author information, bug reporting URL, etc.
+
+1. The mandatory metadata must guarantee _binary compatibility_: allow plugins
```

Comment:
I think this needs to be clearer about the amount of binary and source compatibility we provide.

I also don't think we should break source compatibility outside of major versions, which is fairly easy to do if we always add new API entry points instead of editing / removing existing ones.

Reply:
Done.